File: pset6/readability/readability.py
import string
import re
from cs50 import get_string

def main():

    text = get_string("Text: ")

    letters = len(re.findall('[a-zA-Z]', text))
    words = len(text.split());
    sentences = len(re.findall(r'[.!?]+', text))

    index = readability_index(letters, words, sentences)

    print_grade(index)


def readability_index(letters, words, sentences):

    words_per_100 = words / 100
    l = letters / words_per_100
    s = sentences / words_per_100

    index = round(0.0588 * l - 0.296 * s - 15.8)

    return index


def print_grade(index):
    if index < 1:
        print("Before Grade 1")
    elif index >= 16:
        print("Grade 16+")
    else:
        print(f"Grade {index}")


main()

File: pset2/readability/readability.c
#include <stdio.h>
#include <cs50.h>
#include <string.h>
#include <ctype.h>
#include <math.h>

int count_letters(string text);
int count_words(string text);
int count_sentences(string text);
int readability_index(int letters, int words, int sentences);

int main(void)
{
    string text = get_string("Text: ");
    int letters = count_letters(text);
    int words = count_words(text);
    int sentences = count_sentences(text);
    int index = readability_index(letters, words, sentences);
    if (index < 1)
    {
        printf("Before Grade 1\n");
    }
    else if (index >= 16)
    {
        printf("Grade 16+\n");
    }
    else
    {
        printf("Grade %i\n", index);
    }
}


int count_letters(string text)
{
    int letters = 0;
    for (int i = 0, n = strlen(text); i < n; i++)
        if (isalpha(text[i]))
        {
            letters ++;
        }
    return letters;
}

int count_words(string text)
{
    int words = 0;
    for (int i = 0, n = strlen(text); i < n; i++)
    {
        if (text[i] == '-' && isalpha(text[i - 1])) {}

        else if (text[i] == 39 && isalpha(text[i - 1])) {}

        else if (!isalpha(text[i]) && isalpha(text[i - 1]))
        {
            words ++;
        }

    }
    return words;
}

int count_sentences(string text)
{
    int sentences = 0;
    for (int i = 0, n = strlen(text); i < n; i++)
    {
        int character = text[i];
        if (character == 46 || character == 33 || character == 63)
        {
            sentences ++;
        }
    }
    return sentences;
}

int readability_index(int letters, int words, int sentences)
{
    float words_per_100 = words / 100.0;
    float L = letters / words_per_100;
    float S = sentences / words_per_100;
    int index = round(0.0588 * L - 0.296 * S - 15.8);
    return index;
}